### PR TITLE
Networked terrain

### PR DESCRIPTION
**Pull request by peter-r-g**
Contains tons of terrain things. Networking, optimization, and general cleanup of the terrain.

### CODE REVIEW

File: code/States/Gamemodes/BaseGamemode.cs [hunk 1]
```diff
@@ -24,6 +24,11 @@ public abstract partial class BaseGamemode : BaseState
 	/// </summary>
 	[Net]
 	public TeamManager TeamManager { get; private set; } = null!;
+	/// <summary>
```
Comment (trundler-dev) [inline code/States/Gamemodes/BaseGamemode.cs:27]: I know it's annoying but can we ensure there are empty lines between before a doc comment? (i.e. a space between lines 26 and 27 here)

File: code/States/Gamemodes/BaseGamemode.cs [hunk 2]
```diff
@@ -62,6 +67,12 @@ public abstract partial class BaseGamemode : BaseState
 
 	protected override void Enter( bool forced, params object[] parameters )
 	{
+		if ( !IsServer )
+		{
+			base.Enter( forced, parameters );
+			return;
+		}
+
 		if ( GameConfig.TerrainFile != string.Empty )
```
Comment (trundler-dev) [inline code/States/Gamemodes/BaseGamemode.cs:76]: Perhaps we should move the terrain-related initialization out to another function in the future. Fine for now though.

File: code/Terrain/MarchingSquares.cs [hunk 3]
```diff
@@ -109,33 +85,49 @@ public Model CreateWallModel()
 			}
 		}
 
-		if ( Host.IsClient )
+		_builder.AddCollisionMesh( wallVertices.ToArray(), wallTriangles.ToArray() );
+		// CreateWallModel
+
```
Comment (trundler-dev) [inline code/Terrain/MarchingSquares.cs:90]: Weird spacing for this comment here.

Comment (trundler-dev) [reply code/Terrain/MarchingSquares.cs:90]: And in a few spots down below too.

File: code/States/Gamemodes/BaseGamemode.cs (hunk 1 above)
Comment (trundler-dev) [reply code/States/Gamemodes/BaseGamemode.cs:27]: And anywhere else too. Not a huge deal but just keep in mind for the future. I wonder if there is an editorconfig option for this.

**Review by trundler-dev — APPROVED** [review]
Minor things but overall you can just merge it in. Awesome work!

File: code/Terrain/MarchingSquares.cs (hunk 3 above)
Comment (peter-r-g) [reply code/Terrain/MarchingSquares.cs:90]: I left comments on the old method names there so you can see where your methods went in the merge

File: code/States/Gamemodes/BaseGamemode.cs (hunk 1 above)
Comment (peter-r-g) [reply code/States/Gamemodes/BaseGamemode.cs:27]: I can sort that in master after it's merged

File: code/States/Gamemodes/BaseGamemode.cs (hunk 2 above)
Comment (peter-r-g) [reply code/States/Gamemodes/BaseGamemode.cs:76]: Yeah I can move it to a virtual method like `SetupTerrain` so other game modes can override it

**Pull request merged**